### DISCUSSION TITLE
Implement typescript in Yoga components

### DISCUSSION THREAD

**Discussion by mmartinsoliv** [Gympass/yoga#459] [Ideas]
Hi,

We are thinking of migrating Yoga to typescript, with the justification of improving the experience of those who use the lib and for those who develop and contribute to Yoga.

The initial idea would be to migrate the `components`, `theme`, and the `system` to typescript. Of course, migrating to typescript requires some effort. In conversations with @tcK1 and @nypacheco, we had different but very interesting points and I'd like to collect more feedback on the migration!

**Comment by caiotracera** [comment]
I think it's a great idea!

Do you already have any formed opinions that will make typings available directly in lib or through [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped)? Also, do you have any preference for what will be worked on first? The components, theme or the systems?

**Reply by mmartinsoliv** [reply]
At the moment I don't have any opinion formed. But I think of something like a definition file for example `d.ts` and when importing the library has the option to import the types with npm or yarn.

I believe it would be interesting to start with the theme and the system.

**Reply by mmartinsoliv** [reply]
There is an "automatic" way is with [ts-migrate](https://github.com/airbnb/ts-migrate), however, it only migrates the code and it would be necessary to change the pipeline and tooling to transpile the code

**Comment by evilamaior** [comment]
I would like to understand better the technical motivations behind this and why we would like to implement TS in Yoga. What are the pros and cons of using TS?

Also, do our software engineers know TS? Could this be a barrier to contribution to Yoga? These are some points that I have doubts about and would like to clarify.

In addition, do we have examples of other design systems that use TS? I think it would be nice to study why they chose this language and maybe even do a benchmarking!

**Reply by luispiresgympass** [reply]
I don't know how the Core team is setting their application, but here in New Ventures is an Engineer Standard to use Typescript for all new projects related to Javascript.
In that case, today, we are ignoring Yoga in the `global.d.ts` to not have parse problems related to Yoga and the Types that we don't have.

For some cases, in our application, we started to create some types using Styled Components + Yoga for the tokens, which is very helpful.
Also, for IntelliSense and for developing features without going to the docs to see all the props we have (improve productivity).

Some design system examples in Typescript:
https://github.com/callstack/react-native-paper
https://github.com/mui/material-ui
